Fix issue #1466

ThomasKiesel opened the pull request:
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Is backward-compatible?    | ✔️
| Tests pass?   | ✔️
| Fixed issues  | #1466
| Updated README/docs?   | ❌
| Added CHANGELOG entry?   | ✔️

ThomasKiesel commented:
I do not know how to handle the failing test:

- on my machine, the test succeeds
- I do not see how my change have an influence on this test at all. I changed the code of the array editor, but the tested scheme does not contain an array at all.

Is it possible that the test is a false positive? Can I rerun this test?

schmunk42 commented:
> I do not know how to handle the failing test

@ThomasKiesel Which failing test :) ? All tests pass.
But there is a new test page ... without a test itself, maybe this is missing from the commits?

ThomasKiesel commented:
Great to see all tests passing.
Indeed, we created a test page for examining the problem in the browser developer tools. Unfortunately, we did not find a way to write an automated test (which could be hard for asserting no memory leaks). We still left the test page in the commit that one might be able to observe the issue later manually.
If this is not fine for you, we should rather remove the test page from the code base.

germanbisurgi commented:
I followed the tests page instructions

> Remove Item 2, then destroy editor. The row of Item 2 which is only stored in the cache is not destroyed.

I can't check what's in the row_cache because the editor was deleted and no longer exists (garbage collected).

How can I find out where the memory leak is occurring?

ThomasKiesel commented:
First of all, we have to apologize because we observed the memory leak in an editor with a more complex scheme, and obviously, we boiled down the test too much such that the memory leak is not visible anymore.

We updated the test page: Now, we have an array of objects instead an array of strings. Furthermore, we added a button to reproduce the steps multiple times.

You can make two observations:

1. If you set a breakpoint in the `destroyRow` method below the `if (hard)` condition (such that it is hit only if `hard === true`), you will find that the third row is hidden and kept in `row_cache` when item 2 is destroyed, but when the editor is finally completely destroyed, this row in the cache is never destroyed 'hard' as all other rows, that means `row.destroy()` is never called for it.
2.  You can open the page in Chrome and make an initial memory snapshot. After that, you perform the following steps
- Press "start"
- Remove Item 2
- Press "destroyEditor"
Take a second memory snapshot. Now, repeat the described steps five times and take a third memory snapshot. 
Now you can compare the third memory snapshot with the second. Order the table descendingly by '#Delta' and search for the class `JSONEditor`. We find that five new instances have been created. If you select one of them, you find that they are retained by the `ObjectEditor` of the row. 
If the rows in the cache are destroyed with `hard=true`, these instances are not retained.

Screenshot of my memory profile:
![Screenshot Memory Profile Json Editor](https://github.com/json-editor/json-editor/assets/44029697/bf4fcafd-7556-426d-8b83-2046228b4623)

germanbisurgi commented:
@ThomasKiesel thanks for your time. I can reproduce this following your instructions and can indeed see, that multiple JSONEditor instances are still in memory after destroy. I also added your changes and the leaked instances were no more there.

@schmunk42 I think this is a good fix should be merged.

germanbisurgi commented:
Thank you
